### PR TITLE
CISRV-1758 added destro free to cli

### PR DESCRIPTION
**Pull request by matthewalbani**

### CODE REVIEW

**Comment by matthewalbani** [comment]
This one is less ready then the other two. The cli still will send the test info and the repo config, but the api server will ignore it. I am going to do some clean up and try and make it send less un-needed info.

**Comment by matthewalbani** [comment]
@semipermeable as i am doing some cleanup anything that stands out as needing change to you?

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 
 ## Version History
-
+- v1.30.0 = Added destrofree support
```
Comment (semipermeable) [inline CHANGELOG.md]: This will need more explanation before it's published...

File: lib/solano/cli/api.rb [hunk 2]
```diff
@@ -281,7 +288,37 @@ def get_sessions(params={})
...
+      params.merge!({:v3 => true})
+      call_api(:post, "#{Api::Path::REPO_SNAPSHOT}", params)
+    end
+
+    def request_patch_url(params={})
+      params.merge!({:v3 => true})
+      call_api(:post, "#{Api::Path::SESSION_PATCH}/request_url", params)
+    end
+
+    def uploaed_session_patch(params={})
```
Comment (semipermeable) [inline lib/solano/cli/api.rb]: typo:  s/uploaed/upload/

File: lib/solano/cli/solano.rb [hunk 3]
```diff
@@ -32,11 +32,19 @@ def initialize(*args)
...
-                                                        1, 
-                                                        caller_version, 
+      cli_opts[:debug] = true
+      @tddium_client = TddiumClient::InternalClient.new(options[:host],
+                                                        options[:port],
+                                                        options[:proto],
+                                                        1,
+                                                        caller_version,
+                                                        cli_opts)
+      @cli_options = options
```
Comment (semipermeable) [inline lib/solano/cli/solano.rb]: duplicated with line 49?

File: CHANGELOG.md (hunk 1 above)
Comment (semipermeable) [inline CHANGELOG.md]: In fact, we should version this as pre-release for now so there isn't any confusion.

File: lib/solano/scm/git.rb [hunk 4]
```diff
@@ -129,6 +129,77 @@ def number_of_commits(id_from, id_to)
...
+      #check if commit is known locally
+      if (`git branch -q --contains #{snapshot_commit} 2>&1 >/dev/null | grep -o 'error:' | wc -l`).to_i > 0 then
+        say "Snapshot commit does not exist in current repo"
+        create_snapshot(session_id, options)
+        return
+      end
+      say "Snapshot commit exist in current repo"
+
+      file_name = "patch.diff"
+      file_path = File.expand_path("./#{file_name}")
```
Comment (semipermeable) [inline lib/solano/scm/git.rb]: This should go into `Dir.tmpdir`.

File: lib/solano/scm/git.rb [hunk 5]
```diff
@@ -129,6 +129,77 @@ def number_of_commits(id_from, id_to)
       result.split("\n").length
     end
 
+    def create_snapshot(session_id, options={})
+      api = options[:api]
+      res = api.request_snapshot_url({:session_id => session_id})
+      auth_url = res['auth_url']
+      say "Snapshot url is #{auth_url}"
+
+      snaphot_path = File.expand_path("./.snapshot")
```
Comment (semipermeable) [inline lib/solano/scm/git.rb]: This should go into `Dir.tmpdir`

File: lib/solano/scm/git.rb [hunk 6]
```diff
@@ -129,6 +129,77 @@ def number_of_commits(id_from, id_to)
...
+
+      desc = {"url" => auth_url.gsub(/\?.*/,''),
+        "size" => File.stat(file).size,
+        "sha1"=> Digest::SHA1.file(file).hexdigest.upcase,
+        "commit_id"=> `git rev-parse HEAD`,
+        "session_id" => session_id,
+      }
+      api.update_snapshot({:repo_snapshot => desc})
+
+      FileUtils.rm_rf(snaphot_path)
```
Comment (semipermeable) [inline lib/solano/scm/git.rb]: Cleanup should move into an `ensure` block.

File: lib/solano/scm/git.rb (hunk 4 above)
Comment (semipermeable) [inline lib/solano/scm/git.rb]: And the filename should have a unique id in it, like pid + timestamp.

File: lib/solano/scm/git.rb [hunk 7]
```diff
@@ -129,6 +129,77 @@ def number_of_commits(id_from, id_to)
...
+        say "Snapshot commit does not exist in current repo"
+        create_snapshot(session_id, options)
+        return
+      end
+      say "Snapshot commit exist in current repo"
+
+      file_name = "patch.diff"
+      file_path = File.expand_path("./#{file_name}")
+      out = `git format-patch #{snapshot_commit} --stdout > #{file_name}`
+      say "Created patch.diff"
```
Comment (semipermeable) [inline lib/solano/scm/git.rb]: Messages should move into constants.

File: lib/solano/scm/git.rb [hunk 8]
```diff
@@ -129,6 +129,77 @@ def number_of_commits(id_from, id_to)
...
+    def create_patch(session_id, options={})
+      api = options[:api]
+      snapshot_commit = options[:commit]
+      if "#{snapshot_commit}" == `git rev-parse HEAD`.to_s.strip then
+        say "snapshot commit is the same as HEAD"
+        return
+      end
+      #check if commit is known locally
+      if (`git branch -q --contains #{snapshot_commit} 2>&1 >/dev/null | grep -o 'error:' | wc -l`).to_i > 0 then
+        say "Snapshot commit does not exist in current repo"
```
Comment (semipermeable) [inline lib/solano/scm/git.rb]: I think we discussed having this default to failing out instead of creating snapshots that might compete with each other.

Comment (matthewalbani) [inline lib/solano/scm/git.rb]: @semipermeable so only create a snapshot if there is not one? if the commit id is not found error?

File: CHANGELOG.md (hunk 1 above)
Comment (matthewalbani) [inline CHANGELOG.md]: not sure what you mean here?

File: lib/solano/cli/commands/spec.rb [hunk 9]
```diff
@@ -216,7 +234,7 @@ def spec(*pattern)
 
         # XXX time out if all tests are done and the session isn't done.
         if current_test_executions['session_done'] ||
-          (finished_tests.size >= num_tests_started && (Time.now - last_finish_timestamp) > Default::TEST_FINISH_TIMEOUT)
+          ((manager != 'DestroFreeSessionManager' && finished_tests.size >= num_tests_started) && (Time.now - last_finish_timestamp) > Default::TEST_FINISH_TIMEOUT)
```
Comment (matthewalbani) [inline lib/solano/cli/commands/spec.rb]: change to a phase check

File: lib/solano/scm/git.rb [hunk 10]
```diff
@@ -129,6 +129,74 @@ def number_of_commits(id_from, id_to)
...
+        create_snapshot(session_id, options)
+        return
+      end
+
+      file_name = "#{SecureRandom.hex(10)}.patch"
```
Comment (matthewalbani) [inline lib/solano/scm/git.rb]: add solano

File: lib/solano/scm/git.rb [hunk 11]
```diff
@@ -129,6 +129,74 @@ def number_of_commits(id_from, id_to)
...
 
+    def create_snapshot(session_id, options={})
+      api = options[:api]
+      res = api.request_snapshot_url({:session_id => session_id})
+      auth_url = res['auth_url']
+
+      say Text::Process::SNAPSHOT_URL % auth_url
+
+      snaphot_path = File.join(Dir.tmpdir,".snapshot")
+      file = File.join(Dir.tmpdir, 'snapshot.tar')
```
Comment (matthewalbani) [inline lib/solano/scm/git.rb]: add unique

File: lib/solano/scm/git.rb [hunk 12]
```diff
@@ -129,6 +129,74 @@ def number_of_commits(id_from, id_to)
...
     end
 
+    def create_snapshot(session_id, options={})
+      api = options[:api]
+      res = api.request_snapshot_url({:session_id => session_id})
+      auth_url = res['auth_url']
+
+      say Text::Process::SNAPSHOT_URL % auth_url
+
+      snaphot_path = File.join(Dir.tmpdir,".snapshot")
```
Comment (matthewalbani) [inline lib/solano/scm/git.rb]: also unique

File: lib/solano/scm/git.rb [hunk 13]
```diff
@@ -129,6 +129,74 @@ def number_of_commits(id_from, id_to)
...
+      if (`git branch -q --contains #{snapshot_commit} 2>&1 >/dev/null | grep -o 'error:' | wc -l`).to_i > 0 then
+        say "Snapshot commit does not exist in current repo"
+        create_snapshot(session_id, options)
+        return
+      end
+
+      file_name = "#{SecureRandom.hex(10)}.patch"
+      file_path = File.join(Dir.tmpdir, file_name)
+      out = `git format-patch #{snapshot_commit} --stdout > #{file_path}`
+      say "Created patch.diff"
```
Comment (matthewalbani) [inline lib/solano/scm/git.rb]: this one

File: lib/solano/scm/scm.rb [hunk 14]
```diff
@@ -69,6 +69,37 @@ def push_latest(session_data, suite_details, options={})
       return false
     end
 
+    def create_snapshot(session_id, options={})
+      say "snapshot creation not supported"
```
Comment (matthewalbani) [inline lib/solano/scm/scm.rb]: constant

File: lib/solano/cli/commands/spec.rb [hunk 15]
```diff
@@ -295,11 +336,11 @@ def read_and_encode_latest_commits
 
     def docker_enabled
       if @repo_config['system'] then
-        if @repo_config['system']['docker'] then
-          @repo_config['system']['docker']
-        else
-          false
-        end
+         if @repo_config['system']['docker'] then
```
Comment (bkendzior) [inline lib/solano/cli/commands/spec.rb]: nit - Why three spaces?

**Comment by semipermeable** [comment]
@matthewalbani can you rebase with master?

**Comment by matthewalbani** [comment]
rebased

**Pull request merged**